src/s3path.jl

```diff
@@ -0,0 +1,235 @@
+    # "s3://bucket/a/" returns ["s3://bucket/a/b"].
+    if isempty(fp.segments)
+        key = ""
+    elseif fp.dir
+        key = fp.key
+    else
+        return false
+    end
+
+    objects = s3_list_objects(fp.config, fp.bucket, key; max_items=1)
+    return !isempty(objects)
+end
+
+Base.real(fp::S3Path) = fp
+function FilePathsBase.stat(fp::S3Path)
+    # Currently AWSS3 would require a s3_get_acl call to fetch
+    # ownership and permission settings
+    m = Mode(user=(READ + WRITE), group=(READ + WRITE), other=(READ + WRITE))
+    u = FilePathsBase.User()
+    g = FilePathsBase.Group()
+    blksize = 4096
+    blocks = 0
+    s = 0
+    last_modified = DateTime(0)
+
+    if exists(fp)
+        resp = s3_get_meta(fp.config, fp.bucket, fp.key)
+        # Example: "Thu, 03 Jan 2019 21:09:17 GMT"
+        last_modified = DateTime(
+            resp["Last-Modified"][1:end-4],
+            dateformat"e, d u Y H:M:S",
+        )
+        s = parse(Int, resp["Content-Length"])
+        blocks = ceil(Int, s / 4096)
+    end
+
+    return Status(0, 0, m, 0, u, g, 0, s, blksize, blocks, last_modified, last_modified)
+end
+
+# TODO: FilePathsBase should default to calling stat?
+FilePathsBase.lstat(fp::S3Path) = stat(fp)
+FilePathsBase.mode(fp::S3Path) = stat(fp).mode
+Base.size(fp::S3Path) = stat(fp).size
+FilePathsBase.created(fp::S3Path) = stat(fp).ctime
+FilePathsBase.modified(fp::S3Path) = stat(fp).mtime
+Base.islink(fp::S3Path) = false
+Base.issocket(fp::S3Path) = false
+Base.isfifo(fp::S3Path) = false
+Base.ischardev(fp::S3Path) = false
+Base.isblockdev(fp::S3Path) = false
+Base.ismount(fp::S3Path) = false
+# Need API for accessing object ACL permissions for this to work
+FilePathsBase.isexecutable(fp::S3Path) = false
+Base.isreadable(fp::S3Path) = true
+Base.iswritable(fp::S3Path) = true
+
+function Base.mkdir(fp::S3Path; recursive=false, exist_ok=false)
+    fp.dir || throw(ArgumentError("S3Path folders must end with '/': $fp"))
+
+    if exists(fp)
+        !exist_ok && error("$fp already exists.")
+    else
+        if hasparent(fp) && !exists(parent(fp))
+            if recursive
+                mkdir(parent(fp); recursive=recursive, exist_ok=exist_ok)
+            else
+                error(
+                    "The parent of $fp does not exist. " *
+                    "Pass recursive=true to create it."
+                )
+            end
+        end
+
+        write(fp, "")
+    end
+end
+
+function Base.rm(fp::S3Path; recursive=false, kwargs...)
+    if isdir(fp)
+        files = readpath(fp)
+
+        if recursive
+            for f in files
+                rm(f; recursive=recursive, kwargs...)
+            end
+        elseif length(files) > 0
+            error("S3 path $object is not empty. Use `recursive=true` to delete.")
+        end
+    end
+
+    @debug "delete: $fp"
+    s3_delete(fp.config, fp.bucket, fp.key)
+end
+
+function Base.readdir(fp::S3Path)
+    if isdir(fp)
```

Comment:
Do we need
```suggestion
    if FilePathsBase.isdir(fp)
```
?
or is that the same as `Base.isdir`.
If it is, then I have mild objections to the earlier part where we overloiaded `FilePaths.isdir`
and it actually was changing e `Base.isdir`

Reply:
It's the same as `Base.isdir`. If `FilePathsBase` is overloading `Base.isdir` does it matter?